# This patch makes OpenAL Soft use mkxp-polyfill.cpp instead of POSIX threads and `std::mutex` for portability reasons.

--- a/al/auxeffectslot.cpp
+++ b/al/auxeffectslot.cpp
@@ -27,7 +27,7 @@
 #include <cstdint>
 #include <iterator>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <numeric>
 #include <thread>
 
--- a/al/buffer.cpp
+++ b/al/buffer.cpp
@@ -32,7 +32,7 @@
 #include <iterator>
 #include <limits>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <new>
 #include <numeric>
 #include <stdexcept>
--- a/al/effect.cpp
+++ b/al/effect.cpp
@@ -27,7 +27,7 @@
 #include <cstring>
 #include <iterator>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <new>
 #include <numeric>
 #include <utility>
--- a/al/error.cpp
+++ b/al/error.cpp
@@ -30,7 +30,7 @@
 #include <cstdarg>
 #include <cstdio>
 #include <cstring>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 
 #include "AL/al.h"
 #include "AL/alc.h"
--- a/al/event.cpp
+++ b/al/event.cpp
@@ -8,7 +8,7 @@
 #include <cstring>
 #include <exception>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <new>
 #include <string>
 #include <thread>
--- a/al/filter.cpp
+++ b/al/filter.cpp
@@ -28,7 +28,7 @@
 #include <cstdio>
 #include <iterator>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <new>
 #include <numeric>
 
--- a/al/listener.cpp
+++ b/al/listener.cpp
@@ -23,7 +23,7 @@
 #include "listener.h"
 
 #include <cmath>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 
 #include "AL/al.h"
 #include "AL/alc.h"
--- a/al/source.cpp
+++ b/al/source.cpp
@@ -35,7 +35,7 @@
 #include <iterator>
 #include <limits>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <new>
 #include <numeric>
 #include <stdexcept>
--- a/al/state.cpp
+++ b/al/state.cpp
@@ -24,7 +24,7 @@
 
 #include <atomic>
 #include <cmath>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <stdexcept>
 #include <string>
 
--- a/alc/alc.cpp
+++ b/alc/alc.cpp
@@ -46,7 +46,7 @@
 #include <iterator>
 #include <limits>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <new>
 #include <stddef.h>
 #include <stdexcept>
@@ -956,7 +956,7 @@ std::atomic<ALCenum> LastNullDeviceError{ALC_NO_ERROR};
 bool TrapALCError{false};
 
 /* One-time configuration init control */
-std::once_flag alc_config_once{};
+
 
 /* Flag to specify if alcSuspendContext/alcProcessContext should defer/process
  * updates.
@@ -1370,7 +1370,7 @@ void alc_initconfig(void)
 #endif // ALSOFT_EAX
 }
 inline void InitConfig()
-{ std::call_once(alc_config_once, [](){alc_initconfig();}); }
+{ alc_initconfig(); }
 
 
 /************************************************
--- a/alc/alconfig.cpp
+++ b/alc/alconfig.cpp
@@ -35,7 +35,7 @@
 
 #include <algorithm>
 #include <cstdio>
-#include <string>
+#include "../../src/mkxp-polyfill.h"
 #include <utility>
 
 #include "alfstream.h"
@@ -493,21 +493,21 @@ al::optional<std::string> ConfigValueStr(const char *devName, const char *blockN
 al::optional<int> ConfigValueInt(const char *devName, const char *blockName, const char *keyName)
 {
     if(const char *val{GetConfigValue(devName, blockName, keyName)})
-        return static_cast<int>(std::strtol(val, nullptr, 0));
+        return static_cast<int>(std::stol(val, nullptr, 0));
     return al::nullopt;
 }
 
 al::optional<unsigned int> ConfigValueUInt(const char *devName, const char *blockName, const char *keyName)
 {
     if(const char *val{GetConfigValue(devName, blockName, keyName)})
-        return static_cast<unsigned int>(std::strtoul(val, nullptr, 0));
+        return static_cast<unsigned int>(std::stoul(val, nullptr, 0));
     return al::nullopt;
 }
 
 al::optional<float> ConfigValueFloat(const char *devName, const char *blockName, const char *keyName)
 {
     if(const char *val{GetConfigValue(devName, blockName, keyName)})
-        return std::strtof(val, nullptr);
+        return std::stof(val, nullptr);
     return al::nullopt;
 }
 
--- a/alc/context.h
+++ b/alc/context.h
@@ -3,7 +3,7 @@
 
 #include <atomic>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <stdint.h>
 #include <utility>
 
--- a/alc/device.h
+++ b/alc/device.h
@@ -3,7 +3,7 @@
 
 #include <atomic>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <stdint.h>
 #include <string>
 #include <utility>
--- a/alc/effects/base.h
+++ b/alc/effects/base.h
@@ -2,7 +2,7 @@
 #define EFFECTS_BASE_H
 
 #include "core/effects/base.h"
-
+#include "../../../src/mkxp-polyfill.h"
 
 EffectStateFactory *NullStateFactory_getFactory(void);
 EffectStateFactory *ReverbStateFactory_getFactory(void);
--- a/common/almalloc.cpp
+++ b/common/almalloc.cpp
@@ -11,20 +11,15 @@
 #ifdef HAVE_MALLOC_H
 #include <malloc.h>
 #endif
-
+#include "../../src/mkxp-polyfill.h"
 
 void *al_malloc(size_t alignment, size_t size)
 {
     assert((alignment & (alignment-1)) == 0);
     alignment = std::max(alignment, alignof(std::max_align_t));
 
-#if defined(HAVE_POSIX_MEMALIGN)
-    void *ret{};
-    if(posix_memalign(&ret, alignment, size) == 0)
-        return ret;
-    return nullptr;
-#elif defined(HAVE__ALIGNED_MALLOC)
-    return _aligned_malloc(size, alignment);
+#ifdef MKXPZ_HAVE_ANY_ALIGNED_MALLOC
+    return mkxp_aligned_malloc(alignment, size);
 #else
     size_t total_size{size + alignment-1 + sizeof(void*)};
     void *base{std::malloc(total_size)};
@@ -50,10 +45,8 @@ void *al_calloc(size_t alignment, size_t size)
 
 void al_free(void *ptr) noexcept
 {
-#if defined(HAVE_POSIX_MEMALIGN)
-    std::free(ptr);
-#elif defined(HAVE__ALIGNED_MALLOC)
-    _aligned_free(ptr);
+#ifdef MKXPZ_HAVE_ANY_ALIGNED_MALLOC
+    mkxp_aligned_free(ptr);
 #else
     if(ptr != nullptr)
         std::free(*(static_cast<void**>(ptr) - 1));
--- a/common/threads.cpp
+++ b/common/threads.cpp
@@ -86,47 +86,47 @@ void semaphore::post()
 void semaphore::wait() noexcept
 { WaitForSingleObject(static_cast<HANDLE>(mSem), INFINITE); }
 
-bool semaphore::try_wait() noexcept
-{ return WaitForSingleObject(static_cast<HANDLE>(mSem), 0) == WAIT_OBJECT_0; }
+
+
 
 } // namespace al
 
 #else
 
-#include <pthread.h>
-#ifdef HAVE_PTHREAD_NP_H
-#include <pthread_np.h>
-#endif
+#include "../../src/mkxp-polyfill.h"
+
+
+
 #include <tuple>
 
 namespace {
 
 using setname_t1 = int(*)(const char*);
-using setname_t2 = int(*)(pthread_t, const char*);
-using setname_t3 = void(*)(pthread_t, const char*);
-using setname_t4 = int(*)(pthread_t, const char*, void*);
+using setname_t2 = int(*)(mkxp_thread_t, const char*);
+using setname_t3 = void(*)(mkxp_thread_t, const char*);
+using setname_t4 = int(*)(mkxp_thread_t, const char*, void*);
 
 void setname_caller(setname_t1 func, const char *name)
 { func(name); }
 
 void setname_caller(setname_t2 func, const char *name)
-{ func(pthread_self(), name); }
+{ func(mkxp_thread_self(), name); }
 
 void setname_caller(setname_t3 func, const char *name)
-{ func(pthread_self(), name); }
+{ func(mkxp_thread_self(), name); }
 
 void setname_caller(setname_t4 func, const char *name)
-{ func(pthread_self(), "%s", static_cast<void*>(const_cast<char*>(name))); }
+{ func(mkxp_thread_self(), "%s", static_cast<void*>(const_cast<char*>(name))); }
 
 } // namespace
 
 void althrd_setname(const char *name)
 {
-#if defined(HAVE_PTHREAD_SET_NAME_NP)
-    setname_caller(pthread_set_name_np, name);
-#elif defined(HAVE_PTHREAD_SETNAME_NP)
-    setname_caller(pthread_setname_np, name);
-#endif
+
+
+
+
+
     /* Avoid unused function/parameter warnings. */
     std::ignore = name;
     std::ignore = static_cast<void(*)(setname_t1,const char*)>(&setname_caller);
@@ -155,8 +155,8 @@ void semaphore::post()
 void semaphore::wait() noexcept
 { dispatch_semaphore_wait(mSem, DISPATCH_TIME_FOREVER); }
 
-bool semaphore::try_wait() noexcept
-{ return dispatch_semaphore_wait(mSem, DISPATCH_TIME_NOW) == 0; }
+
+
 
 } // namespace al
 
@@ -168,27 +168,27 @@ namespace al {
 
 semaphore::semaphore(unsigned int initial)
 {
-    if(sem_init(&mSem, 0, initial) != 0)
-        throw std::system_error(std::make_error_code(std::errc::resource_unavailable_try_again));
+    if(mkxp_sem_init(&mSem, initial) != 0)
+        abort();
 }
 
 semaphore::~semaphore()
-{ sem_destroy(&mSem); }
+{ mkxp_sem_destroy(&mSem); }
 
 void semaphore::post()
 {
-    if(sem_post(&mSem) != 0)
-        throw std::system_error(std::make_error_code(std::errc::value_too_large));
+    if(mkxp_sem_post(&mSem) != 0)
+        abort();
 }
 
 void semaphore::wait() noexcept
 {
-    while(sem_wait(&mSem) == -1 && errno == EINTR) {
+    while(mkxp_sem_wait(&mSem) == -1 && errno == EINTR) {
     }
 }
 
-bool semaphore::try_wait() noexcept
-{ return sem_trywait(&mSem) == 0; }
+
+
 
 } // namespace al
 
--- a/common/threads.h
+++ b/common/threads.h
@@ -14,7 +14,7 @@
 #if defined(__APPLE__)
 #include <dispatch/dispatch.h>
 #elif !defined(_WIN32)
-#include <semaphore.h>
+#include "../../src/mkxp-polyfill.h"
 #endif
 
 void althrd_setname(const char *name);
@@ -27,7 +27,7 @@ class semaphore {
 #elif defined(__APPLE__)
     using native_type = dispatch_semaphore_t;
 #else
-    using native_type = sem_t;
+    using native_type = mkxp_sem_t;
 #endif
     native_type mSem;
 
@@ -40,7 +40,7 @@ public:
 
     void post();
     void wait() noexcept;
-    bool try_wait() noexcept;
+
 };
 
 } // namespace al
--- a/core/ambdec.cpp
+++ b/core/ambdec.cpp
@@ -16,7 +16,7 @@
 #include "alfstream.h"
 #include "alspan.h"
 #include "opthelpers.h"
-
+#include "../../src/mkxp-polyfill.h"
 
 namespace {
 
--- a/core/dbus_wrap.cpp
+++ b/core/dbus_wrap.cpp
@@ -5,7 +5,7 @@
 
 #ifdef HAVE_DYNLOAD
 
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <type_traits>
 
 #include "logging.h"
--- a/core/dbus_wrap.h
+++ b/core/dbus_wrap.h
@@ -9,7 +9,7 @@
 
 #ifdef HAVE_DYNLOAD
 
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 
 #define DBUS_FUNCTIONS(MAGIC) \
 MAGIC(dbus_error_init) \
--- a/core/device.h
+++ b/core/device.h
@@ -8,7 +8,7 @@
 #include <bitset>
 #include <chrono>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <string>
 
 #include "almalloc.h"
--- a/core/except.cpp
+++ b/core/except.cpp
@@ -7,7 +7,7 @@
 #include <cstdarg>
 
 #include "opthelpers.h"
-
+#include "../../src/mkxp-polyfill.h"
 
 namespace al {
 
--- a/core/helpers.cpp
+++ b/core/helpers.cpp
@@ -9,7 +9,7 @@
 #include <cstdlib>
 #include <cstdio>
 #include <cstring>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <limits>
 #include <string>
 #include <tuple>
--- a/core/hrtf.cpp
+++ b/core/hrtf.cpp
@@ -14,7 +14,7 @@
 #include <fstream>
 #include <iterator>
 #include <memory>
-#include <mutex>
+#include "../../src/mkxp-polyfill.h"
 #include <numeric>
 #include <type_traits>
 #include <utility>
--- a/core/logging.cpp
+++ b/core/logging.cpp
@@ -10,7 +10,7 @@
 #include "alspan.h"
 #include "strutils.h"
 #include "vector.h"
-
+#include "../../src/mkxp-polyfill.h"
 
 #if defined(_WIN32)
 #define WIN32_LEAN_AND_MEAN
--- a/core/mastering.cpp
+++ b/core/mastering.cpp
@@ -15,7 +15,7 @@
 #include "alnumeric.h"
 #include "alspan.h"
 #include "opthelpers.h"
-
+#include "../../src/mkxp-polyfill.h"
 
 /* These structures assume BufferLineSize is a power of 2. */
 static_assert((BufferLineSize & (BufferLineSize-1)) == 0, "BufferLineSize is not a power of 2");
